---
title: "create_annotation"
author: "Victor Yuan"
date: "02/08/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

Outline

1. Get illumina provided cpg and genome locations
2. Use annotatr to provide 'basic' annotations, such as cpg island, gene elements, enhacners
3. Use published datasets to provide other interesting annotation info (e.g. imprinting)
4. Use rtracklayer::liftOver to map from build hg19 to build hg38 genome locations

# Libraries 

```{r}
library(rtracklayer) 
library(annotatr)
library(tidyverse)
library(readxl)
library(fuzzyjoin)
library(janitor)
```

# 1. Illumina annotation

We start with the illumina provided locations of each probe. Note that Illumina has released multiple versions of the EPIC, each with a different set of probes (200-2000 probes can be different between arrays).

 Here I combine the b4 annotation with those probes that were removed from B2 to B3 versions. This set of probes
will be a starting point for cpgs to annotate.

TODO: Add in b2 and b1 annotations to ensure all EPIC probes in the universe will be covered.
TODO: Add in 450k probes?
TODO: Add in 27k probes?

```{r}
b4 <- read_csv('Z:/Victor/Data/DNAm annotations/MethylationEPIC_v-1-0_B4.csv', skip = 7)
b4_missing <- read_csv('Z:/Victor/Data/DNAm annotations/MethylationEPIC Missing Legacy CpG (v1.0_B3 vs. v1.0_B2) Annotations.csv')
anno <- bind_rows(b4, b4_missing %>% mutate_at(c('AddressA_ID', 'AddressB_ID'), as.character))

# clean up annotation
anno <- anno %>% mutate(Strand = ifelse(Strand == 'F', '+', '-'),
                start = MAPINFO,
                end = MAPINFO,
                chr = paste0('chr', CHR),
                cpg_id = Name) %>%
  
  # keep basic information
  dplyr::select(cpg_id, chr, start, end, Strand, AddressA_ID:Forward_Sequence) %>%
  
  # remove rs, and control probes
  filter(grepl('^[(cg)(ch)]{1}', cpg_id)) %>%
  
  # add illumina identifier
  rename_at(vars(-(1:5)), ~paste0('ilmn_', .))
```

# 2. Use annotatr to add UCSC annotations, and FANTOM5 enhancers

The gather and spread functions take 10 hours to complete.

First I filter to the illumina information to keep. 

Except for Start, End, chr and cpg_id columns, I rename others with the prefix 'ilmn', to indicate 
they are from the illumina annotation file. People can refer to the illumina files for description 
of those variables

```{r}
# coerce to granges
anno_GR <- makeGRangesFromDataFrame(anno, keep.extra.columns = T)

# select annotations to use
annots <- c('hg19_enhancers_fantom',
            'hg19_cpgs', 
            'hg19_genes_1to5kb',
            'hg19_genes_promoters', 
            'hg19_genes_5UTRs',
            'hg19_genes_exons',
            'hg19_genes_intronexonboundaries',
            'hg19_genes_introns',
            'hg19_genes_3UTRs',
            'hg19_genes_intergenic')
annots <- build_annotations(genome = 'hg19', annotations = annots)

# annotate the base annotation
annotated <- annotate_regions(regions = anno_GR, annotations = annots, ignore.strand = T) %>% 
  as_tibble

# clean it up
annotated
x1 <- length(unique(annotated$cpg_id))

system.time(
x <- annotated %>% 
  
  # 1. remove identifiers from annot.id, i.e "intron.9999" should be just called "intron"
  mutate(annot.id = str_extract(annot.id, '^[A-z0-9]+'),
         annot.type = str_extract(annot.type, '(?<=hg19_)[^_]+')
         ) %>%
  dplyr::select(-annot.seqnames, -annot.start, -annot.end, -annot.strand) %>%
  
  # 2. paste together all elements, grouped by cpg-related and gene-related elements
  ### I'm not sure but this step might be able to skip
  group_by(cpg_id,annot.type) %>%
  summarize(
    # paste(unique(.)) ensures all nonunique mappings are retained
    cpg = paste(unique(cpg_id)),
    chr = paste(unique(seqnames)),
    start = paste(unique(start)),
    end = paste(unique(end)),
    
    id = paste(annot.id, collapse = ", "),
    width = paste(annot.width, collapse = ", "),
    tx_id = paste(annot.tx_id, collapse = ", "),
    gene_id = paste(annot.gene_id, collapse = ", "),
    symbol = paste(annot.symbol, collapse = ", "))  %>%
  
  # 3. now spread multiple values across columns
  # first create a temporary variable to spread by, using gather() and unite()
  ungroup() %>%
  dplyr::select(-cpg_id)  %>%
  gather(variable, value, -(annot.type:end)) %>%
  unite(temp, annot.type, variable, sep = '_') %>%
  spread(temp, value) %>%
  
  # 4. clean up
  # remove uninformative columns
  dplyr::select(-enhancers_gene_id , -enhancers_symbol, -enhancers_tx_id, 
                -cpg_gene_id, -cpg_symbol, -cpg_tx_id) %>%
  
  # cpg_id == inter replace with "sea"
  mutate(cpg_id = ifelse(cpg_id == 'inter', 'sea', cpg_id)) %>%
  mutate_all(~replace(., . == 'NA', NA_character_))
) # takes 10 hours
# ensure the number of cpgs ended up with equals the number of unique cpgs starting out with
x1 == nrow(x) # must be T 

# now add back in the illumina columns
  
```
saveRDS(x, '../../data/main/interim/annotation.rds')
write_csv(x, '../../data/main/processed/annotation.csv')

x <- readRDS('Z:/Victor/Projects/NIH - cells/data/main/interim/annotation.rds')

# 3. Add other annotations

PMDs

https://www.pnas.org/content/110/15/6037

Schroeder DI, Blair JD, Lott P, Yu HOK, Hong D, Crary F, et al. The human placenta methylome. PNAS. 2013 Apr 9;110(15):6037â€“42. 

```{r}
# load pmds
pmds <- read_xls('Z:/Victor/Projects/NIH - cells/data/external/placental_pmds.xls', col_names = F)
colnames(pmds) <- c('chr', 'start', 'end')

# join pmds using fuzzyjoin::genome_left_join
anno <- x %>%
  mutate_at(c('start', 'end'), as.numeric) %>%
  genome_left_join(pmds) %>%
  
  # pmd ID is the genome location
  mutate(pmd_id = ifelse(!is.na(chr.y), paste0(chr.y, ':', start.y, '-', end.y), NA),
         pmd_width = end.y - start.y) %>%
  dplyr::rename(chr = chr.x, start = start.x, end = end.x) %>%
  select(-contains('.y'))
```

Imprinting

List of imprinting regions were created by Giulia 

```{r}
imprint <- read_xlsx('Imprinted gene database.xlsx', col_names = T, sheet = 'Imprinted genes and DMRs')%>%
  clean_names() %>%
  mutate(in_gene = ifelse(grepl('Gene', location_type), 'Gene', NA),
         in_dmr = ifelse(grepl('DMR', location_type), 'DMR', NA))

# imprinted genes
imprint_gene <- imprint  %>%
  select(gene, gene_location, placenta_specific, in_gene) %>% 
  separate(gene_location, into = c('chr', 'start', 'end'), convert = T)  %>%
  
  # create imprint_gene columns to join to anno
  dplyr::rename(imprinted_gene = gene) %>% 
  filter(!is.na(imprinted_gene)) %>%
  mutate(imprinted_gene_general = ifelse(is.na(placenta_specific), imprinted_gene, NA),
         imprinted_gene_placenta = ifelse(!is.na(placenta_specific), imprinted_gene, NA)) 

anno <- anno %>% 
  genome_left_join(imprint_gene %>% select(-imprinted_gene, -placenta_specific, -in_gene)) %>%
  dplyr::rename(chr = chr.x, start = start.x, end = end.x) %>%
  select(-contains('.y'))

# duplicate rows for multiples genes, combine into single rows as output
anno_x <- anno %>% 
  group_by(cpg) %>% 
  summarize(imprinted_gene_general = paste(imprinted_gene_general, collapse = ';'))

anno <- anno %>% select(-imprinted_gene_general) %>% distinct() %>%
  left_join(anno_x) %>%
  mutate(imprinted_gene_general = ifelse(imprinted_gene_general == 'NA', NA, imprinted_gene_general))


# imprinted dmrs
imprint_dmr <- imprint %>%
  filter(in_dmr == 'DMR') %>%
  
  # input: multiple dmrs per row, output: single dmr per row 
  select(dmr_location, placenta_specific) %>%
  separate(dmr_location, into = c('dmr_location_1', 'dmr_location_2', 'dmr_location_3', 
                                  'dmr_location_4', 'dmr_location_5'), 
           sep = '[\\s]*[;][\\s]*', fill = 'right') %>%
  gather(key = 'key', value = 'dmr_location', dmr_location_1:dmr_location_5,
         na.rm = T) %>%
  select(-key) %>%
  
  # clean, make new indicator columns for dmrs
  mutate(dmr_location = gsub(' ', '', dmr_location),
         imprinted_dmr_general = ifelse(is.na(placenta_specific), dmr_location, NA),
         imprinted_dmr_placenta = ifelse(!is.na(placenta_specific), dmr_location, NA))  %>%
  separate(dmr_location, into = c('chr', 'start', 'end'), convert = T)  %>%
  select(-placenta_specific)
  
anno <- anno %>% 
  genome_left_join(imprint_dmr) %>%
  dplyr::rename(chr = chr.x, start = start.x, end = end.x) %>%
  select(-contains('.y')) %>%
  
  # rearrange columns
  clean_names %>%
  select(cpg:end, strand, cpg_id:genes_width, pmd_id:imprinted_dmr_placenta, contains('ilmn'))
```

# 4. Map from hg19 to hg38

I provide a mapping from hg19 to hg38 coordinates. Note that this loses 247 cpgs.

```{r eval = T}
# load chain file
chain <- import.chain("hg19ToHg38.over.chain")
chain

# coerce to granges object
anno_gr38 <- anno %>%
  makeGRangesFromDataFrame(keep.extra.columns = T)
seqlevelsStyle(anno_gr38) <- 'UCSC'

# hg19 to hg38 coordinates
anno_gr38 <- liftOver(anno_gr38, chain) %>% unlist

nrow(anno) - length(anno_gr38) # loses 247 loci
setdiff(anno$cpg, anno_gr38$cpg) # these specific loci

# clean
anno_gr38 <- anno_gr38 %>% as_tibble %>%
  dplyr::rename(chr = seqnames) %>%
  select(cpg, everything())

```

# Save data

```{r eval = F}
saveRDS(anno, file = 'hg19_epic_annotation.rds')
saveRDS(anno_gr38, file = 'hg38_epic_annotation.rds')
```

# 2019-09-27 Update

PMD coordinates were actually in hg18 coordinates. So here I convert them to hg19/gr38 before replacing into the annotations

```{r}
anno <- readRDS(file = 'hg19_epic_annotation.rds')
anno_gr38 <- readRDS(file = 'hg38_epic_annotation.rds')
```

## liftover

```{r}
pmds <- read_xls('Z:/Victor/Projects/NIH - cells/data/external/placental_pmds.xls', col_names = F)
colnames(pmds) <- c('chr', 'start', 'end')

# make an identifier for original pmds
pmds <- pmds %>% 
  mutate(width = end-start,
         ID = paste0('pmd_', 1:n()))

# load chain file
chain_hg18_hg19 <- import.chain("hg18ToHg19.over.chain")
chain_hg18_hg19


# coerce to granges object
pmds_hg18 <- pmds %>%
  makeGRangesFromDataFrame(keep.extra.columns = T)

# hg18 to hg19 coordinates
pmds_hg19 <- liftOver(pmds_hg18, chain_hg18_hg19) %>% unlist

length(pmds_hg19) - length(pmds_hg18) # gains 4681 ranges
setdiff(pmds_hg18$ID, pmds_hg19$ID) # lost 1 pmd


pmds_hg19 %>%
  as_tibble() %>%
  group_by(ID) %>%
  filter(n() > 1) %>%
  left_join(pmds, by = 'ID')
```

Note that by lifting over, many of the pmds have been "broken" up. It seems that in hg19 many sequences were added into the genome. So a pmd is broken up when there is a new sequence added somewhere into the middle of it.

Should I take the remapped hg19 pmds as is?
Or should I say take the boundary for each original pmd in it's new coordinates, and merge broken pieces back together?

For now, I leave the remapped pmds as is.

```{r}
# hg18 to gr38

# load chain file
chain_hg18_gr38 <- import.chain("hg18ToHg38.over.chain")
chain_hg18_gr38

pmds_gr38 <- liftOver(pmds_hg18, chain_hg18_gr38) %>% unlist

length(pmds_gr38) - length(pmds_hg18) # gains 43355 ranges
setdiff(pmds_hg18$ID, pmds_gr38$ID) # lost 6 pmds
```

## Add to annotation

```{r}
#hg19
pmds_hg19 <- pmds_hg19 %>%
  as_tibble() %>% 
  rename(chr = seqnames,
         pmd_width = width) %>%
  mutate(pmd_id = paste0(chr, ':', start, '-', end)) %>%
  dplyr::select(-ID, -strand)

anno <- anno %>%
  dplyr::select(-pmd_id, -pmd_width) %>%
  genome_left_join(pmds_hg19, by = c('chr', 'start', 'end')) %>%
  rename(chr = chr.x, start = start.x, end = end.x) %>%
  dplyr::select(-contains('.y'))


# gr38
pmds_gr38 <- pmds_gr38 %>%
  as_tibble() %>% 
  rename(chr = seqnames,
         pmd_width = width) %>%
  mutate(pmd_id = paste0(chr, ':', start, '-', end)) %>%
  dplyr::select(-ID, -strand)

anno_gr38 <- anno_gr38 %>%
  dplyr::select(-pmd_id, -pmd_width) %>%
  genome_left_join(pmds_gr38, by = c('chr', 'start', 'end')) %>%
  rename(chr = chr.x, start = start.x, end = end.x) %>%
  dplyr::select(-contains('.y'))
```

## Save data

```{r eval = F}
saveRDS(anno, file = 'hg19_epic_annotation.rds')
saveRDS(anno_gr38, file = 'hg38_epic_annotation.rds')
```

# 2019-11-05 Update

Noticed there are some NAs in the genes_symbol column

```{r, eval = FALSE}
anno <- readRDS(file = 'hg19_epic_annotation.rds')
anno_gr38 <- readRDS(file = 'hg38_epic_annotation.rds')

anno %>% 
  mutate(genes_symbol = gsub(', NA', '', 
                             gsub('NA ,', '', genes_symbol))) %>%
  saveRDS(file = 'hg19_epic_annotation.rds')

anno_gr38 %>% 
  mutate(genes_symbol = gsub(', NA', '', 
                             gsub('NA ,', '', genes_symbol))) %>%
  saveRDS(file = 'hg19_epic_annotation.rds')
```

# Column descriptions

```{r}
## column descriptors
colnames(x)
# cpg: cpg identifier
# chr: chromosome position
# start, end: cpg position in bp 
# cpg_id: relation to cpg island features. uses UCSC cpg islands as reference, then definitions for shelves, shores, and seas are used
# cpg_width: width in bp of the cpg island -related feature
# enhancers_id: fantom5 enhancer
# enhancers_width: width in bp of the fantom5 enhancer
# genes_: columns are retrieved from txdb.hsapiens.ucsc.hg19.knowngene
# genes_gene_id: I'm not sure what this is TBH
# genes_id: gene-related elemtn, e.g. promoters, introns/exons, based off of UCSC
# genes_symbol: Gene symbol
# genes_tx_id: transcript identifier
# genes_width: width in bp of gene element
# pmd_id, pmd_width: indicates which pmd a given cpg lies in. list taken from d.i.schoreder et al. 2013
# imprinted_gene_: indicates the imprinted gene a given cpg lies in, placental or non-tissue specific
# imprinted_dmr_: indicates the imprinted dmr a given cpg lies in, placental or non-tissue specific
# imprinting regions/genes combined from several databases and primary articles, process by GDG.
# columns prefixed by 'ilmn_' come from the illumina provided annotation files
```


