---
title: "Annotating DNA methylation array"
date: "07/08/2019"
output:
    html_document:
      keep_md: yes
      theme: spacelab
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: no
editor_options: 
  chunk_output_type: console
---

```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE)
```

This repository contains code for generating an annotation for the Illumina EPIC methylation array.

There are two annotation files, one mapped to **hg19** and one mapped to **hg38/GR38**. They have the same
annotation information (columns), but the hg38 annotation is missing 237 probes, since some mappings
are lost from converting from hg19 to hg38.

```{r include = F, message = F, warning = F }
library(tidyverse)
library(kableExtra)
```

```{r, message = F, warning = F}
anno <- readRDS(file = 'hg19_epic_annotation.rds')
anno_gr38 <- readRDS(file = 'hg38_epic_annotation.rds')
```


# The process

## 1. Starting annotation

I started with the default annotations provided by Illumina. I used two files, the latest b4 annotation (*MethylationEPIC_v-1-0_B4.csv*), and the list of probes that are missing between b3 and b2 (*MethylationEPIC Missing Legacy CpG (v1.0_B3 vs. v1.0_B2) Annotations.csv*). Both can be found on the product files list from [Illumina's website](https://support.illumina.com/downloads/infinium-methylationepic-v1-0-product-files.html).

Using the intersection of these two lists of probes, I used the provided genomic location (chromomsome and position) to map annotations to each cpg. Note that Illumina's provided annotations are based on hg19.

```{r}
anno %>% select(cpg, chr, start) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) 
```

I also kept some probe-specific information that I thought some may find useful. The columns for these variables are all prefixed with "ilmn_".

```{r}
anno %>% select(cpg, chr, start, contains('ilmn')) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

## 2. Annotate cpgs

I used the R package annotatr to access UCSC annotations for cpg islands and transcripts, and then FANTOM5 for enhancers.

### UCSC transcript and cpg island -related elements:

```{r}
anno %>% select(cpg, chr, start, cpg_id:cpg_width, genes_id:genes_width) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

### Enhancers

```{r}
anno %>% select(cpg, chr, start, contains('enhancer')) %>%
  filter(!is.na(enhancers_id)) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

### PMDs from Schroeder et al. 2013:

Taken from the primary article.

```{r}
anno %>% select(cpg, chr, start, contains('pmd')) %>%
  filter(!is.na(pmd_id)) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

### Imprinting regions

Both general and placenta specific from GeneImprint and OTAGO databases, plus a court et al. and hanna et al. DMR imprinting papers.

These imprinting information was summarized and provided by GDG

```{r}
anno %>% select(cpg, chr, start, contains('imprint')) %>%
  filter(!is.na(imprinted_gene_placenta) | !is.na(imprinted_gene_general)) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

## 3. Map to hg38

Lastly I mapped the annotation to the genome assembly hg38 using UCSC liftover's tool implemented in R.

