---
title: "Annotating DNA methylation array"
date: "07/08/2019"
output:
    html_document:
      keep_md: yes
      theme: spacelab
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: no
editor_options: 
  chunk_output_type: console
---

```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE)
```

This repository contains code for generating an annotation for the Illumina EPIC methylation array.

There are two annotation files, one mapped to **hg19** (`hg19_epic_annotation.rds`) and one mapped to **hg38/GR38** (`hg38_epic_annotation.rds`). They have the same annotation information (columns), but the hg38 annotation is missing 237 probes, since some mappings are lost from converting from hg19 to hg38.

Currently these files are not git tracked because they are too large (~250 mb).

```{r include = F, message = F, warning = F }
library(tidyverse)
library(kableExtra)
```

```{r, message = F, warning = F}
anno <- readRDS(file = 'hg19_epic_annotation.rds')
```

# The process

## 1. Starting annotation

I started with the default annotations provided by Illumina. I used two files, the latest b4 annotation (*MethylationEPIC_v-1-0_B4.csv*), and the list of probes that are missing between b3 and b2 (*MethylationEPIC Missing Legacy CpG (v1.0_B3 vs. v1.0_B2) Annotations.csv*). Both can be found on the product files list from [Illumina's website](https://support.illumina.com/downloads/infinium-methylationepic-v1-0-product-files.html).

Using the intersection of these two lists of probes, I used the provided genomic location (chromomsome and position) to map annotations to each cpg. Note that Illumina's provided annotations are based on hg19.

> an example of the starting coordinates from Illumina that this annotation is based on

```{r}
anno %>% select(cpg, chr, start) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) 
```

I also kept some probe-specific information that I thought some may find useful. The columns for these variables are all prefixed with "ilmn_".

```{r}
anno %>% select(cpg, chr, start, contains('ilmn')) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

## 2. Annotate cpgs

### Transcript-related features, enhancers, cpg islands

I used the R package `annotatr` to access UCSC annotations for cpg islands / transcripts, and FANTOM5 for enhancers.

> UCSC transcript and cpg island -related elements:

```{r}
anno %>% select(cpg, chr, start, cpg_id:cpg_width, genes_id:genes_width) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

> Enhancers

```{r}
anno %>% select(cpg, chr, start, contains('enhancer')) %>%
  filter(!is.na(enhancers_id)) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

### Placental partially methylated domains (PMDs) from Schroeder et al. 2013:

Taken from the primary article.

```{r}
anno %>% select(cpg, chr, start, contains('pmd')) %>%
  filter(!is.na(pmd_id)) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

### Imprinting regions

These placental imprinted regions were collected from several sources. The merging of these regions into a combined resource is documented at [github.com/wvictor14/human_methylation_imprints](github.com/wvictor14/human_methylation_imprints).

```{r}
anno %>% 
  select(cpg, chr, start, contains('imprint')) %>%
  filter(!is.na(imprint_tissue_specificity) | !is.na(imprint_methylated_allele)) %>%
  dplyr::slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = '100%')
```

## 3. Map to hg38

Lastly I mapped the annotation to the genome assembly hg38 using UCSC liftover's tool implemented in R. This results in a loss of 237 cpgs.

